### PR TITLE
Base implementation

### PR DESCRIPTION
**Pull request by Maar-io**
Adding Base functionality for RMRK.
Base enables equipping functionality which will be covered in next PR.

This ink! implementation follows evm implementation of Base defined [here](https://github1s.com/rmrk-team/evm/blob/master/contracts/RMRK/base/IRMRKBaseStorage.sol)
Implemented functions are:

- [x] add_part_list()
- [x] add_equippable_addresses()
- [x] reset_equippable_addresses()
- [x] set_equippable_by_all()
- [x] setup_base()

Getters
- [x] get_base_metadata()
- [x] get_parts_count()
- [x] get_part()
- [x] is_equippable()
- [x] is_equippable_by_all()

Events

- [ ] No events are implemented. Unclear at this point if Events for Base are needed.

Test
- [x]  Unit test
- [x]  e2e test with typechain (test/base.spec.ts)

### CODE REVIEW

File: logics/impls/rmrk/base.rs [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+    T: Storage<BaseData> + Storage<ownable::Data>,
+{
+    /// Add one or more parts to the base
+    #[modifiers(only_owner)]
+    default fn add_part_list(&mut self, parts: Vec<Part>) -> Result<(), PSP34Error> {
+        for part in parts {
+            let part_id = self.data::<BaseData>().next_part_id;
+
+            if part.part_type == PartType::Fixed {
+                if part.equippable.len() != 0 || part.is_equippable_by_all {
```
Comment (PierreOssun) [inline logics/impls/rmrk/base.rs:59]: These 2 if's can be one

File: tests/base.spec.ts [hunk 2]
```diff
@@ -0,0 +1,139 @@
...
+  let dave: KeyringPair;
+  let kanaria: Rmrk;
+  let gem: Rmrk;
+
+  const ZERO_ADDRESS = encodeAddress(
+    '0x0000000000000000000000000000000000000000000000000000000000000000',
+  );
+  let gasRequired: bigint;
+
+  async function setup(): Promise<void> {
```
Comment (PierreOssun) [inline tests/base.spec.ts:42]: you can use BeforeEach

**Review by PierreOssun — APPROVED** [review]
LGTM
left some clean code comments

File: logics/impls/rmrk/base.rs [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+    /// Add collection address(es) that can be used to equip given `PartId`.
+    #[modifiers(only_owner)]
+    default fn add_equippable_addresses(
+        &mut self,
+        part_id: PartId,
+        equippable_address: Vec<AccountId>,
+    ) -> Result<(), PSP34Error> {
+        let mut part = self.ensure_only_slot(part_id)?;
+        for address in equippable_address {
```
[the remark this answers is not in the document]
Comment (Maar-io) [reply logics/impls/rmrk/base.rs:81]: great!!! I was trying with append but it did not like it.

File: tests/base.spec.ts (hunk 2 above)
Comment (Maar-io) [reply tests/base.spec.ts:42]: Do you have any example for async BeforeEach?
Created an issue for now #21

File: logics/impls/rmrk/base.rs (hunk 1 above)
Comment (Maar-io) [reply logics/impls/rmrk/base.rs]: done

**Pull request merged**